foreign key constraint failed to add on clean install.

Completely clean install of MySQL 8.12, Tomcat 9.0 and gitbucket 4.28 on Windows Server 2012R2.

Gitbucket fails to start after partially initialising the tables in MySQL. The error in Tomcat log is as follows:

24-Sep-2018 15:11:31.997 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [gitbucket.core.servlet.InitializeListener]
 java.sql.SQLException: (conn=89) Failed to add the foreign key constraint. Missing index for constraint 'IDX_ISSUE_COMMENT_FK0' in the referenced table 'ISSUE'
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:198)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:110)
	at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:228)
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:216)
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.execute(MariaDbPreparedStatementClient.java:150)
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeUpdate(MariaDbPreparedStatementClient.java:183)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at io.github.gitbucket.solidbase.migration.MigrationUtils.updateDatabase(MigrationUtils.java:20)
	at io.github.gitbucket.solidbase.migration.LiquibaseMigration.migrate(LiquibaseMigration.java:93)
	at io.github.gitbucket.solidbase.migration.LiquibaseMigration.migrate(LiquibaseMigration.java:54)
	at io.github.gitbucket.solidbase.Solidbase.migrate(Solidbase.java:45)
	at io.github.gitbucket.solidbase.Solidbase.migrate(Solidbase.java:29)
	at gitbucket.core.servlet.InitializeListener.$anonfun$contextInitialized$1(InitializeListener.scala:69)
	at gitbucket.core.servlet.InitializeListener.$anonfun$contextInitialized$1$adapted(InitializeListener.scala:59)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$2(BlockingProfile.scala:207)
	at slick.JdbcProfileBlockingSession$BlockingSession.withTransaction(TransactionalJdbcBackend.scala:26)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$1(BlockingProfile.scala:207)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withSession(BlockingProfile.scala:200)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withTransaction(BlockingProfile.scala:207)
	at gitbucket.core.servlet.InitializeListener.contextInitialized(InitializeListener.scala:59)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4643)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5109)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:743)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:719)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:703)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:986)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1858)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.util.concurrent.AbstractExecutorService.submit(Unknown Source)
	at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:772)
	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:426)
	at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1585)
	at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:308)
	at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
	at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:424)
	at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:367)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:969)
	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:839)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1429)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1419)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.util.concurrent.AbstractExecutorService.submit(Unknown Source)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:944)
	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:261)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:422)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:770)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.startup.Catalina.start(Catalina.java:682)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:350)
	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:492)
Caused by: java.sql.SQLException: Failed to add the foreign key constraint. Missing index for constraint 'IDX_ISSUE_COMMENT_FK0' in the referenced table 'ISSUE'
Query is: ALTER TABLE gitbucket.ISSUE_COMMENT ADD CONSTRAINT IDX_ISSUE_COMMENT_FK0 FOREIGN KEY (USER_NAME, REPOSITORY_NAME, ISSUE_ID) REFERENCES gitbucket.ISSUE (USER_NAME, REPOSITORY_NAME, ISSUE_ID)
	at org.mariadb.jdbc.internal.util.LogQueryTool.exceptionWithQuery(LogQueryTool.java:153)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:255)
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:209)
	... 58 more

Any help would be appeciated.
Kind Regards,
--Colin

--- a/build.sbt
+++ b/build.sbt
@@ -66,8 +66,7 @@ libraryDependencies ++= Seq(
   "junit"                           % "junit"                        % "4.12" % "test",
   "org.scalatra"                    %% "scalatra-scalatest"          % ScalatraVersion % "test",
   "org.mockito"                     % "mockito-core"                 % "2.23.4" % "test",
-  "com.wix"                         % "wix-embedded-mysql"           % "4.2.0" % "test",
-  "ru.yandex.qatools.embed"         % "postgresql-embedded"          % "2.10" % "test",
+  "com.dimafeng"                    %% "testcontainers-scala"        % "0.22.0" % "test",
   "net.i2p.crypto"                  % "eddsa"                        % "0.3.0",
   "is.tagomor.woothee"              % "woothee-java"                 % "1.8.0",
   "org.ec4j.core"                   % "ec4j-core"                    % "0.0.3"

--- a/build.sbt
+++ b/build.sbt
@@ -67,6 +67,8 @@ libraryDependencies ++= Seq(
   "org.scalatra"                    %% "scalatra-scalatest"          % ScalatraVersion % "test",
   "org.mockito"                     % "mockito-core"                 % "2.23.4" % "test",
   "com.dimafeng"                    %% "testcontainers-scala"        % "0.22.0" % "test",
+  "org.testcontainers"              % "mysql"                        % "1.10.3" % "test",
+  "org.testcontainers"              % "postgresql"                   % "1.10.3" % "test",
   "net.i2p.crypto"                  % "eddsa"                        % "0.3.0",
   "is.tagomor.woothee"              % "woothee-java"                 % "1.8.0",
   "org.ec4j.core"                   % "ec4j-core"                    % "0.0.3"

--- a/src/main/resources/update/gitbucket-core_4.0.xml
+++ b/src/main/resources/update/gitbucket-core_4.0.xml
@@ -206,7 +206,7 @@
     <column name="PULL_REQUEST" type="boolean" nullable="false"/>
   </createTable>
 
-  <addPrimaryKey constraintName="IDX_ISSUE_PK" tableName="ISSUE" columnNames="ISSUE_ID, USER_NAME, REPOSITORY_NAME"/>
+  <addPrimaryKey constraintName="IDX_ISSUE_PK" tableName="ISSUE" columnNames="USER_NAME, REPOSITORY_NAME, ISSUE_ID"/>
   <addForeignKeyConstraint constraintName="IDX_ISSUE_FK0" baseTableName="ISSUE" baseColumnNames="USER_NAME, REPOSITORY_NAME" referencedTableName="REPOSITORY" referencedColumnNames="USER_NAME, REPOSITORY_NAME"/>
   <addForeignKeyConstraint constraintName="IDX_ISSUE_FK2" baseTableName="ISSUE" baseColumnNames="MILESTONE_ID" referencedTableName="MILESTONE" referencedColumnNames="MILESTONE_ID"/>
   <addForeignKeyConstraint constraintName="IDX_ISSUE_FK1" baseTableName="ISSUE" baseColumnNames="OPENED_USER_NAME" referencedTableName="ACCOUNT" referencedColumnNames="USER_NAME"/>

--- a/src/test/scala/gitbucket/core/GitBucketCoreModuleSpec.scala
+++ b/src/test/scala/gitbucket/core/GitBucketCoreModuleSpec.scala
@@ -1,17 +1,13 @@
 package gitbucket.core
 
 import java.sql.DriverManager
-import java.time.Duration
-import java.time.temporal.ChronoUnit.SECONDS
 
-import com.dimafeng.testcontainers.GenericContainer
+import com.dimafeng.testcontainers.{MySQLContainer, PostgreSQLContainer}
 import io.github.gitbucket.solidbase.Solidbase
 import io.github.gitbucket.solidbase.model.Module
 import liquibase.database.core.{H2Database, MySQLDatabase, PostgresDatabase}
 import org.junit.runner.Description
 import org.scalatest.{FunSuite, Tag}
-import org.testcontainers.containers.ContainerLaunchException
-import org.testcontainers.containers.wait.strategy.{HostPortWaitStrategy, Wait}
 
 object ExternalDBTest extends Tag("ExternalDBTest")
 
@@ -28,104 +24,44 @@ class GitBucketCoreModuleSpec extends FunSuite {
 
   implicit private val suiteDescription = Description.createSuiteDescription(getClass)
 
-  test("Migration MySQL 5.7", ExternalDBTest) {
-    val container = GenericContainer(
-      "mysql:5.7",
-      env = Map("MYSQL_ROOT_PASSWORD" -> "my-secret-pw", "MYSQL_DATABASE" -> "gitbucket"),
-      waitStrategy = new HostPortWaitStrategy {
-        override def waitUntilReady(): Unit = {
-          super.waitUntilReady()
-
-          def readyForConnections(retry: Int = 0): Boolean = {
-            var con: java.sql.Connection = null
-            try {
-              con = DriverManager.getConnection(
-                s"jdbc:mysql://${waitStrategyTarget.getContainerIpAddress}:${waitStrategyTarget.getMappedPort(3306)}/gitbucket?useSSL=false",
-                "root",
-                "my-secret-pw"
-              )
-              con.createStatement().execute("SELECT 1")
-            } catch {
-              case _: Exception if retry < 3 =>
-                Thread.sleep(10000)
-                readyForConnections(retry + 1)
-              case _: Exception => false
-            } finally {
-              Option(con).foreach(_.close())
-            }
-          }
-
-          if (!readyForConnections()) throw new ContainerLaunchException("Timed out")
+  Seq("8.0", "5.7").foreach { tag =>
+    test(s"Migration MySQL $tag", ExternalDBTest) {
+      val container = new MySQLContainer() {
+        override val container = new org.testcontainers.containers.MySQLContainer(s"mysql:$tag") {
+          override def getDriverClassName = "org.mariadb.jdbc.Driver"
         }
+        // TODO https://github.com/testcontainers/testcontainers-java/issues/736
+        container.withCommand("mysqld --default-authentication-plugin=mysql_native_password")
       }
-    )
-
-    container.starting()
-    try {
-      new Solidbase().migrate(
-        DriverManager.getConnection(
-          s"jdbc:mysql://${container.containerIpAddress}:${container.mappedPort(3306)}/gitbucket?useSSL=false",
-          "root",
-          "my-secret-pw"
-        ),
-        Thread.currentThread().getContextClassLoader(),
-        new MySQLDatabase(),
-        new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
-      )
-    } finally {
-      container.finished()
+      container.starting()
+      try {
+        new Solidbase().migrate(
+          DriverManager.getConnection(s"${container.jdbcUrl}?useSSL=false", container.username, container.password),
+          Thread.currentThread().getContextClassLoader(),
+          new MySQLDatabase(),
+          new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
+        )
+      } finally {
+        container.finished()
+      }
     }
   }
 
-  test("Migration PostgreSQL 11", ExternalDBTest) {
-    val container = GenericContainer(
-      "postgres:11",
-      env = Map("POSTGRES_PASSWORD" -> "mysecretpassword", "POSTGRES_DB" -> "gitbucket"),
-      waitStrategy = Wait
-        .forLogMessage(".*database system is ready to accept connections.*\\s", 2)
-        .withStartupTimeout(Duration.of(60, SECONDS))
-    )
+  Seq("11", "10").foreach { tag =>
+    test(s"Migration PostgreSQL $tag", ExternalDBTest) {
+      val container = PostgreSQLContainer(s"postgres:$tag")
 
-    container.starting()
-    try {
-      new Solidbase().migrate(
-        DriverManager.getConnection(
-          s"jdbc:postgresql://${container.containerIpAddress}:${container.mappedPort(5432)}/gitbucket",
-          "postgres",
-          "mysecretpassword"
-        ),
-        Thread.currentThread().getContextClassLoader(),
-        new PostgresDatabase(),
-        new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
-      )
-    } finally {
-      container.finished()
-    }
-  }
-
-  test("Migration PostgreSQL 10", ExternalDBTest) {
-    val container = GenericContainer(
-      "postgres:10",
-      env = Map("POSTGRES_PASSWORD" -> "mysecretpassword", "POSTGRES_DB" -> "gitbucket"),
-      waitStrategy = Wait
-        .forLogMessage(".*database system is ready to accept connections.*\\s", 2)
-        .withStartupTimeout(Duration.of(60, SECONDS))
-    )
-
-    container.starting()
-    try {
-      new Solidbase().migrate(
-        DriverManager.getConnection(
-          s"jdbc:postgresql://${container.containerIpAddress}:${container.mappedPort(5432)}/gitbucket",
-          "postgres",
-          "mysecretpassword"
-        ),
-        Thread.currentThread().getContextClassLoader(),
-        new PostgresDatabase(),
-        new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
-      )
-    } finally {
-      container.finished()
+      container.starting()
+      try {
+        new Solidbase().migrate(
+          DriverManager.getConnection(container.jdbcUrl, container.username, container.password),
+          Thread.currentThread().getContextClassLoader(),
+          new PostgresDatabase(),
+          new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
+        )
+      } finally {
+        container.finished()
+      }
     }
   }
 

--- a/src/test/scala/gitbucket/core/GitBucketCoreModuleSpec.scala
+++ b/src/test/scala/gitbucket/core/GitBucketCoreModuleSpec.scala
@@ -1,19 +1,17 @@
 package gitbucket.core
 
 import java.sql.DriverManager
+import java.time.Duration
+import java.time.temporal.ChronoUnit.SECONDS
 
+import com.dimafeng.testcontainers.GenericContainer
 import io.github.gitbucket.solidbase.Solidbase
 import io.github.gitbucket.solidbase.model.Module
 import liquibase.database.core.{H2Database, MySQLDatabase, PostgresDatabase}
+import org.junit.runner.Description
 import org.scalatest.{FunSuite, Tag}
-import com.wix.mysql.EmbeddedMysql._
-import com.wix.mysql.config.Charset
-import com.wix.mysql.config.MysqldConfig._
-import com.wix.mysql.distribution.Version._
-import ru.yandex.qatools.embed.postgresql.PostgresStarter
-import ru.yandex.qatools.embed.postgresql.config.AbstractPostgresConfig.{Credentials, Net, Storage, Timeout}
-import ru.yandex.qatools.embed.postgresql.config.PostgresConfig
-import ru.yandex.qatools.embed.postgresql.distribution.Version.Main.PRODUCTION
+import org.testcontainers.containers.ContainerLaunchException
+import org.testcontainers.containers.wait.strategy.{HostPortWaitStrategy, Wait}
 
 object ExternalDBTest extends Tag("ExternalDBTest")
 
@@ -28,52 +26,106 @@ class GitBucketCoreModuleSpec extends FunSuite {
     )
   }
 
-  test("Migration MySQL", ExternalDBTest) {
-    val config = aMysqldConfig(v5_7_latest)
-      .withPort(3306)
-      .withUser("sa", "sa")
-      .withCharset(Charset.UTF8)
-      .withServerVariable("bind-address", "127.0.0.1")
-      .build()
+  implicit private val suiteDescription = Description.createSuiteDescription(getClass)
 
-    val mysqld = anEmbeddedMysql(config)
-      .addSchema("gitbucket")
-      .start()
+  test("Migration MySQL 5.7", ExternalDBTest) {
+    val container = GenericContainer(
+      "mysql:5.7",
+      env = Map("MYSQL_ROOT_PASSWORD" -> "my-secret-pw", "MYSQL_DATABASE" -> "gitbucket"),
+      waitStrategy = new HostPortWaitStrategy {
+        override def waitUntilReady(): Unit = {
+          super.waitUntilReady()
 
+          def readyForConnections(retry: Int = 0): Boolean = {
+            var con: java.sql.Connection = null
+            try {
+              con = DriverManager.getConnection(
+                s"jdbc:mysql://${waitStrategyTarget.getContainerIpAddress}:${waitStrategyTarget.getMappedPort(3306)}/gitbucket?useSSL=false",
+                "root",
+                "my-secret-pw"
+              )
+              con.createStatement().execute("SELECT 1")
+            } catch {
+              case _: Exception if retry < 3 =>
+                Thread.sleep(10000)
+                readyForConnections(retry + 1)
+              case _: Exception => false
+            } finally {
+              Option(con).foreach(_.close())
+            }
+          }
+
+          if (!readyForConnections()) throw new ContainerLaunchException("Timed out")
+        }
+      }
+    )
+
+    container.starting()
     try {
       new Solidbase().migrate(
-        DriverManager.getConnection("jdbc:mysql://localhost:3306/gitbucket?useSSL=false", "sa", "sa"),
+        DriverManager.getConnection(
+          s"jdbc:mysql://${container.containerIpAddress}:${container.mappedPort(3306)}/gitbucket?useSSL=false",
+          "root",
+          "my-secret-pw"
+        ),
         Thread.currentThread().getContextClassLoader(),
         new MySQLDatabase(),
         new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
       )
     } finally {
-      mysqld.stop()
+      container.finished()
     }
   }
 
-  test("Migration PostgreSQL", ExternalDBTest) {
-    val runtime = PostgresStarter.getDefaultInstance()
-    val config = new PostgresConfig(
-      PRODUCTION,
-      new Net("localhost", 5432),
-      new Storage("gitbucket"),
-      new Timeout(),
-      new Credentials("sa", "sa")
+  test("Migration PostgreSQL 11", ExternalDBTest) {
+    val container = GenericContainer(
+      "postgres:11",
+      env = Map("POSTGRES_PASSWORD" -> "mysecretpassword", "POSTGRES_DB" -> "gitbucket"),
+      waitStrategy = Wait
+        .forLogMessage(".*database system is ready to accept connections.*\\s", 2)
+        .withStartupTimeout(Duration.of(60, SECONDS))
     )
 
-    val exec = runtime.prepare(config)
-    val process = exec.start()
-
+    container.starting()
     try {
       new Solidbase().migrate(
-        DriverManager.getConnection("jdbc:postgresql://localhost:5432/gitbucket", "sa", "sa"),
+        DriverManager.getConnection(
+          s"jdbc:postgresql://${container.containerIpAddress}:${container.mappedPort(5432)}/gitbucket",
+          "postgres",
+          "mysecretpassword"
+        ),
         Thread.currentThread().getContextClassLoader(),
         new PostgresDatabase(),
         new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
       )
     } finally {
-      process.stop()
+      container.finished()
+    }
+  }
+
+  test("Migration PostgreSQL 10", ExternalDBTest) {
+    val container = GenericContainer(
+      "postgres:10",
+      env = Map("POSTGRES_PASSWORD" -> "mysecretpassword", "POSTGRES_DB" -> "gitbucket"),
+      waitStrategy = Wait
+        .forLogMessage(".*database system is ready to accept connections.*\\s", 2)
+        .withStartupTimeout(Duration.of(60, SECONDS))
+    )
+
+    container.starting()
+    try {
+      new Solidbase().migrate(
+        DriverManager.getConnection(
+          s"jdbc:postgresql://${container.containerIpAddress}:${container.mappedPort(5432)}/gitbucket",
+          "postgres",
+          "mysecretpassword"
+        ),
+        Thread.currentThread().getContextClassLoader(),
+        new PostgresDatabase(),
+        new Module(GitBucketCoreModule.getModuleId, GitBucketCoreModule.getVersions)
+      )
+    } finally {
+      container.finished()
     }
   }
 

